Allow tron to run after unclean shutdown

kawaiwanyelp opened the pull request:
### Description
- In the event of an unclean shutdown, Tron's pidfile and manhole socket could be left behind in the working directory, preventing another Tron daemon running, despite there being no existing daemon. However, these changes will allow us ignore those orphaned files and start Tron regardless.
- `tron.pid`:
    - If we find a pid file and verify no process is running with that pid, instead of exiting, we now simply clear it out and continue on.
    - We do not delete it, as we have already acquired a lock on it.
- `manhole.sock`:
    - If `manhole.sock.lock` was also left behind after an unclean shutdown, Tron is already able to reclaim the lock and start with no problems.
    - However, if the lock file was removed but the socket file remains, Tron would previously refuse to start. Now, the Tron daemon will simply delete the old manhole socket and continue on.

### Testing
- manual testing with `make dev`
- `make test`

### Notes to reviewers
- I also moved the call to `setup_logging` to the `TronDaemon` initialization because otherwise we wouldn't be able to properly log before actually running the daemon, like, say, during setup.

keymone commented:
> In the event of an unclean shutdown, Tron's pidfile and manhole socket could be left behind in the working directory, preventing another Tron daemon running, despite there being no existing daemon. However, these changes will allow us ignore those orphaned files and start Tron regardless.

does it make sense having pidfiles at all then?

kawaiwanyelp commented:
@keymone, we want to keep pidfiles as a way of ensuring only one Tron daemon is running. If one is left behind, we can use it to check if another daemon is running; we only start the daemon once we verify that another is not running.

qui commented:
I think we should get rid of the PID file entirely, and modify the flock file to implement the behavior we want. Killing whatever process ID is in the PID file is possibly dangerous because PIDs can be reused, and it's unnecessary.

kawaiwanyelp commented:
I've replaced most the pidfile stuff with lockfile stuff everywhere; that includes the docs. However, I've still kept some pidfile CLI options code around for the purposes of backwards compatibility. Once we've updated our environment variables for Tron (which includes a PIDFILE but not LOCKFILE variable), we can fully remove the pidfile code.

keymone commented on debian/tron.service:
```diff
@@ -5,7 +5,7 @@ After=network.target
 User=tron
 EnvironmentFile=/etc/default/tron
-ExecStart=/usr/bin/trond --pid-file=${PIDFILE} --working-dir=${WORKINGDIR} -H 0.0.0.0 -P 8089 -l ${LOGCONF}
+ExecStart=/usr/bin/trond --lock-file=${LOCKFILE:-$PIDFILE} --working-dir=${WORKING+DIR} -H 0.0.0.0 -P 8089 -l ${LOGCONF}
```
`WORKING+DIR` <- that seems wrong?

kawaiwanyelp replied:
Oops!

keymone commented on tron/trondaemon.py:
```diff
@@ -26,65 +26,43 @@
+    """ Create and maintain a lock for the daemon """
 
     def __init__(self, filename):
         self.lock = flockfile.FlockFile(filename)
```
i know this can seem like neverending task, but we're depending on 6 year old wrapper of a library deprecated 3 years ago.

can we ditch all of that and just use stdlib's fctnl?

docs: https://docs.python.org/3/library/fcntl.html#fcntl.flock

example: http://tilde.town/~cristo/file-locking-in-python.html

kawaiwanyelp replied:
Yeah, I can do that.

qui commented on tron/trondaemon.py:
```diff
@@ -160,30 +140,36 @@ def __init__(self, options):
+        # This condition is made with the assumption that no existing daemon
+        # is running. If there is one, the following code could potentially
+        # cause problems for the other daemon by removing its socket.
+        if (os.path.exists(self.manhole_sock) and
```
Hm, this is confusing. What about getting rid of wantPID below, so no lock file is created? We don't really need one anymore. Then you can always delete the manhole_sock file if it exists.

kawaiwanyelp replied:
Ah, `wantPID=1` is what's causing the lock to be generated. I'll remove it to avoid confusion.

kawaiwanyelp commented on tron/trondaemon.py:
```diff
@@ -26,65 +26,43 @@
         try:
-            os.unlink(self.filename)
-            log.info(f"Removed pidfile: {self.filename}")
+            os.remove(self.filename)
```
Okay.

qui commented on tests/utils/flockfile_test.py:
```diff
@@ -0,0 +1,68 @@
+
+@mock.patch('fcntl.flock', mock.Mock(side_effect=BlockingIOError), autospec=None)
+def test_release_other_process_locked(mock_flockfile):
+    mock_flockfile._has_lock = True  # not possible, but allows us to test below
```
Same here, what about just asserting that an error is raised, if you attempt to release the lock when something else has locked it already?
